src/Analyzers/shared/FeatureDetection/ProjectFeatureDetector.cs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.LanguageServices;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Threading;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.AspNetCore.Analyzers.FeatureDetection
+{
+    // Be very careful making changes to this file. No project in this repo builds it.
+    //
+    // If you need to verify a change, make a local project (net472) and copy in everything included by this project.
+    //
+    // You'll also need some nuget packages like:
+    // - Microsoft.VisualStudio.LanguageServices
+    // - Microsoft.VisualStudio.Shell.15.0
+    // - Microsoft.VisualStudio.Threading
+    [Export(typeof(ProjectFeatureDetector))]
+    internal class ProjectFeatureDetector
+    {
+        private readonly Lazy<VisualStudioWorkspace> _workspace;
+
+        [ImportingConstructor]
+        public ProjectFeatureDetector(Lazy<VisualStudioWorkspace> workspace)
+        {
+            _workspace = workspace;
+        }
+
+        public async Task<IImmutableSet<string>> DetectFeaturesAsync(string projectFullPath, CancellationToken cancellationToken = default)
```

Comment:
>DetectFeaturesAsync [](start = 49, length = 19)

The flow now is: we call this DetectFeaturesAsync when dependencies are refreshed on Publish overview page. It works fine , however there is a scenario when publish overview page stays open and user does not do any action that would trigger dependencies refresh. This way if he adds SignalR in code we would never be notified. I think at this time it is OK, this suggestion should not be really "real time", they will see it next time the open publish page then. However if we want to address this, your another approach with having real roslyn analyzer can help, since we can wait when Roslyn calls our analyzer on the code change, analyzer in turn notify ProjectFeatureDetector and that one can ping tooling etc..

Reply:
Sure. I'd like to start simple and see if that comes up via feedback. It's something that can be done, but it will require many more moving parts. It also dictates a different API shape, since we now need to have events.